### ISSUE TITLE
Idea: new plugin for hierarchical navigating notes using internal links (Linkfolder)

### ISSUE THREAD

**Issue by VKondakoff** [vrtmrz/obsidian-tagfolder#18]
Hi!

Thanks a lot for Tagfolder: I'm using this plugin daily on all of my devices for hierarchical navigation notes using tags. And it works just great. But sometimes creating notes hierarchy using internal links is more efficient, the using tags.
What about creating new plugin (say, Linkfolder) to hierarchical navigate notes using internal links instead of tags? The logic is pretty similar to Tagfolder: every unique link is an entry in a tree of an internal links and works like a folder, which contains all the notes, that contain that link. And a single note may reside in several places (depending of the links it contains).
This is a quick description, let's discuss this in more detail! What do you think?

**Comment by VKondakoff** [comment]
Hey, @vrtmrz !

Sorry, for bothering you again, but I have unsuccessfully tried to contact you with email, so I'm hoping to draw your attention here.

Please, let's discuss a possibility to create new plugin, rather similar to TagFolder, but which will use internal links as virtual folders.

**Example:** here it is how my notes hierarchy is setup:

<img width="266" alt="Снимок экрана 2023-08-08 в 19 17 49" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/07559603-8e4a-406a-918f-4bd90a5eaef4">

The links (which are mainly the link to the parent note) may be defined in the Frontmatter (Properties) or in the body of the note:

<img width="441" alt="Снимок экрана 2023-08-08 в 19 20 00" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/30549b02-b43a-4018-8a1f-c580f0a6911a">

<img width="441" alt="Снимок экрана 2023-08-08 в 19 20 49" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/b9f67722-f460-42a0-9207-6594ea3aa760">

As the proposed plugin uses _exactly_ the same logic as TagFolder, maybe it will be not so difficult to implement one?
Please, let's discuss this. It seems there is a real need of a plugin, which can visualise notes in hierarchy view using internal links as a sorting source...

Thanks! :)

**Comment by vrtmrz** [comment]
Hello! I am sorry for missing this, and I am disappointed with the extremely hysteric behaviour of my mailer.

I think that this is a good idea! As you mentioned, using outgoing links instead of tags may not be so difficult.
However, it looks like we should make sure of minor things. May I ask for some details?

Referring to the screenshot, `Jazz` has not been seen on the top level. Could we place `Jazz` in between `hardware` and `medicine`? Or is something specific behaviour required?
(In other words, should this be handled as a couple of nested tags?)

The second thing is a bit more conceptual. In this case, `Music` is a folder, but also a note. I am wondering how to express this thing. E.g., which behaviour - open the file, or toggle the folder opened - is suitable when clicking the name.
(If the virtual child (Which has the same name) is acceptable, it could be the same behaviour).
May I ask your opinion, please?

**Comment by VKondakoff** [comment]
Hey, I'm _really_ glad you show some interest in implementing this! I was kind of obsessed with the idea of using links as folders for more than a year... ;) Yes, of course, the idea should be thoroughly discussed first to make sure there are no stupid conceptual flows which may create difficulties in a future.

> Referring to the screenshot, Jazz has not been seen on the top level. Could we place Jazz in between hardware and medicine? Or is something specific behaviour required?
> (In other words, should this be handled as a couple of nested tags?)

The hierarchical scheme on a screenshot is quite simplified. Yes, I think `Jazz` should appear in several places in hierarchy.
I'd like to think of my internal links like a _parent_ relationship. So there is a `Main` note in my vault which is kind of _central_ hub. Several other hub-notes (`Music`, `Jazz`, `Hardware`, `Medicine`) and some notes which are not connected  to other spheres of interests are referring to main by linking to it (`[[Main]]`).
<img width="1086" alt="Снимок экрана 2023-08-09 в 09 22 52" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/9ecc1572-3008-4a06-a52b-cae342ba3672">
In addition to `Main` hub `Jazz` is connected to `Music` and `VKondakoff`, so we can see the hierarchy like this:

- **Main**
    - **Hardware**
        - Hardware notes
  - **Jazz**
       - Jazz notes
  - **Medicine**
       - Medicine notes
  - **Music**
       - **Jazz**
           - Jazz notes
       - Music notes
  - **VKondakoff**
      - **Jazz**
          - Jazz notes
      - VKondakoff notes 
  - Main notes

This looks quite  similar to TagFolder logic.

> The second thing is a bit more conceptual. In this case, Music is a folder, but also a note. I am wondering how to express this thing. E.g., which behaviour - open the file, or toggle the folder opened - is suitable when clicking the name.
> (If the virtual child (Which has the same name) is acceptable, it could be the same behaviour).
> May I ask your opinion, please?
<img width="221" alt="Снимок экрана 2023-08-09 в 09 46 58" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/9e26b197-6ac1-403a-b544-22d88f4d40f1">
Good question! I'm pretty sure that clicking on a note name should open the file regardless of the fact is this a "plain" note or a "folder" note. But all folder notes should have a collapse/expand sign (`>`) next to them. And clicking on this sign should toggle the folder opened.

**Comment by VKondakoff** [comment]
A couple of months ago I created a sample vault with some linked files (for testing reasons). 
Here it is: https://www.dropbox.com/scl/fi/jxhlhohpgb0jttxu0yhxi/Test.zip?rlkey=s8bakcz13c5al3bpknae4lqgj&dl=0

<img width="484" alt="Снимок экрана 2023-08-09 в 10 06 51" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/f7b6d2bc-a146-4e0c-93d7-02b39beb2329">

And I duplicated all the internal links with tags and here is how they are looking in Tagfolder pane (surprisingly similar to what I want with links):
<img width="1091" alt="Снимок экрана 2023-05-07 в 16 17 08" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/b724bd08-7fa2-4896-845c-66f608a772af">

I wish I can help you somehow with implementing this plugin! At least I'm always ready for some discussion and extensive testing. I can be contacted using [email](mailto:vkondakoff@gmail.com) if this is more convenient to you.

Thanks for your interest!

**Comment by vrtmrz** [comment]
I really appreciate detailed your explanation of the idea! `Links as tags` is an impressibly simple and smart idea.
Even though it was for a PoC, only a small patch was needed to make your idea a reality. Of course, minor cases would have to be checked, however, it could be an optional feature instead of creating a new plugin.

Before we build this up, could you have a quick look at this stage to see if it has as you imagined it, please? We can replace the `main.js` of TagFolder with [this one](https://gist.githubusercontent.com/vrtmrz/a73694fa964c106c9d5ecac2f9813833/raw/088b07651f74608b45f5d6367f7a19d0c1471a39/main.js) to check the behaviour.


Note: In the settings, only `Do not simplify empty folders` is ignored, others work as same as TagFolder.

**Comment by VKondakoff** [comment]
Hey, that's great! Must to say: as a PoC this version works brilliant. I tried it on my main vault with many links here it is how it looks like: 
<img width="447" alt="Снимок экрана 2023-08-10 в 12 01 49" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/2f62dbd4-3585-4bc7-94c2-a50aa3266628">

There are some issues (of course!) which will need some thoughts/discussions from our side. Here are some issues I see after a short time of using the new `main.js`:

- (most important): There are many inline links from the notes which are counted as virtual folders with one file in them which overcrowds the virtual folders list. _Maybe_ we should have a way to filter them in the future (an option, for example, not to show virtual folders when there is, say, less than two notes in this folder)? Or, possible, an option to use only specific links from Properties (frontmatter) to create virtual folders hierarchy (links from `links:` property - this is what I use, or links from the `parent:` property)?
- (second one): The `Expand limit` does not works well now. If I change `Expand limit` to something other than `Level 1` all the available folders are displayed as a subfolders in every folder:
<img width="447" alt="Снимок экрана 2023-08-10 в 12 21 42" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/3b705e64-244d-421e-9bb1-f97b2424f5d1">

- The folder list is not auto-updated after you insert new link (the new note appears after you restart Obsidian). Here is a screenshot (I added `Travel` link to the note Travel and was expecting it to appear in the Travel virtual folder, which happened only after Obsidian restart):
<img width="1070" alt="Снимок экрана 2023-08-10 в 11 27 42" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/bab0ef5c-28a9-43c1-88a2-d4472e821b9c">

It will be really great if `Links as tags` will be implemented in TagFolder in a future: so we can see virtual folders of tags and links in one place. Quick suggestion: to distinguish between link and tags virtual folders use `#`-signs in tags-folders and <ins>underlines</ins> in link-folders.

I will try the new `main.js` on a small test vault I have sent to you yesterday and write back in around two hours!

Thanks a lot!

**Comment by VKondakoff** [comment]
Possible bug: virtual folder name gets truncated from the beginning after I several times switch the `Expand limit` values:

<img width="1032" alt="Снимок экрана 2023-08-10 в 13 26 04" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/b925ca18-7c1d-4a48-a4d2-1b71fe9a9d53">

Maybe this happens when virtual folder gets selected accordingly to a currently open note?

**Comment by VKondakoff** [comment]
New info (the `Expand limit` is set to `No Limit`):

1. Enabling the `Do not treat nested tags as dedicated levels` adds `.md` extension to the links name in the tree-view. After this clicking on a link does nothing:

<img width="452" alt="Снимок экрана 2023-08-11 в 14 18 20" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/200f10b0-36ce-4bb0-8a2c-ecc05b1cb3af">

2. Enabling `Merge redundant combinations` removes the _recursive subfolders_ issue when `Expand limit` is set to something other than `Level 1` I reported earlier:

<img width="450" alt="Снимок экрана 2023-08-11 в 14 24 11" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/1c8e4286-5b16-4171-bad2-145020a3ca5f">

3. Adding `Do not treat nested tags as dedicated levels` to `Merge redundant combinations` gives even better results (and the links are not clickable, unfortunately):

<img width="450" alt="Снимок экрана 2023-08-11 в 14 25 21" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/3faef2b5-2aea-4f9f-b073-a2850df329c0">

Right now the options from `Arrangements` section behave quite unpredictable to me. ;)

**Comment by VKondakoff** [comment]
After more than two weeks of using `Links as tags` I must say that it is _very_ useful even in the current state:

<img width="449" alt="Снимок экрана 2023-08-24 в 22 11 59" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/60c65ac2-41cb-45f5-ae6b-7095c11bb02c">

There are several bugs here (mentioned above) and currently `Links as tags` does not display virtual *sub*folders well, but I'm quite happy to use it just as a plain list of all the available links (containing linked notes).

Great job! Thank you, @vrtmrz, for working on this. Looking forward to see `Links as tags` as dedicated plugin, or (possible) integrated into `TagFolder`. ;)

Thanks again!

**Comment by vrtmrz** [comment]
Thank you for your very detailed review! It really helps me! And I would express gratitude for waiting during my holiday! 
Please excuse the brevity of my response! (Still, I have to unpack or handle the pended jobs).

`Link Folder` has been implemented as an alpha version. We can open it from the command palette; by tapping `Show Link Folder`.

As you mentioned, many links are shown. Property reference mode might be hope.

Other things have been fixed, (or implemented in other ways). However, some configurations (e.g., `Arrangements`, `Filters` ) should be separated. 
We have to work a bit more. But for now, could you please try the v0.17.0?

**Comment by VKondakoff** [comment]
Glad you are back! (Hope you had a great vacation!)

> Link Folder has been implemented as an alpha version. 

The visual representation is just perfect:

<img width="448" alt="Снимок экрана 2023-08-25 в 13 39 14" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/b01cbb95-e35b-4c99-82c7-be5345bc03e0">

**Proposal:** let's think about combining the TagFolder and LinkFolder views in the future - in a single VirtualFolder pane. Pros and cons?

**Issue:** looks like in this alpha (v0.17.0) you use **both** incoming and outgoing links to create virtual folder. This shows _every_ note in my vault as virtual folder:

<img width="1054" alt="Снимок экрана 2023-08-25 в 13 45 23" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/c15c3b26-2fe2-43d0-9a85-5bcf4027b11f">

Here is a note called `Lillet` with one and only link to `Drinks`. I'm expecting to see `Lillet` inside `Drinks` virtual folder (and it is actually there!), but `Lillet` is shown in the tree as a separate subfolder with two children: `Drinks` and `Lillet` itself.
Maybe virtual folder creation should be limited to incoming links only (thinking outgoing link as an `up` or `parent` relationship)?

> As you mentioned, many links are shown. Property reference mode might be hope.

Yes, I see following possible solutions here:
- Have a option to limit LinkFolder to links from Properties only?
- Heve an option to ignore links if there are less than, say, two linked file?
- Have an option to display links with more than, say, two linked files bold or underlined?

Thanks for this alpha!

**Comment by vrtmrz** [comment]
Thank you for your review again!
And also thanks to you all, we had a very good holiday!

Naming: 
Perhaps VirtualFolder, MetaFolder or some good naming is suitable for this after the LinkFolder is stable. However, we do not want to surprise all the familiar people who have been using it for a long time, like you. 
Fortunately, we have a bit more time to think about it.

Direction of Links:
You right! I have merged incoming and outgoing links without deep thought. However, thanks to this, we can control both incoming and outgoing toggling the buttons.

It could be useful. If you have enough time, give it a try, please! (Toggling takes a bit of time).
Also, we can now control the display of "notes linked through several notes".
These are implemented at v0.17.1 and v0.17.2.

> Have a option to limit LinkFolder to links from Properties only?
> Heve an option to ignore links if there are less than, say, two linked file?
> Have an option to display links with more than, say, two linked files bold or underlined?

Perhaps the first one is a suitable and explicit behaviour. Referring to your screenshot, the `Properties` that we are waiting for might help to edit the link in frontmatter.
(Yes, I am a catalyst, but I am waiting for a stable version to shoot issues).
However, a minimal referenced/reference count would also be useful. I will address it in a while!

**Comment by VKondakoff** [comment]
> You right! I have merged incoming and outgoing links without deep thought. However, thanks to this, we can control both incoming and outgoing toggling the buttons.

Thanks, new builds are working much better.
There is one issue: the state of the Direction of links button is not persistent after Obsidian restart. Here is an example.
This is a default state:

<img width="449" alt="Снимок экрана 2023-08-26 в 18 55 21" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/57fd3ce8-03dd-49b4-9e0d-b44577606e8c">

I don't want to see outgoing links, so I toggle the second button:

<img width="449" alt="Снимок экрана 2023-08-26 в 18 55 37" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/63469d80-b702-4c8b-8614-e19972befff8">

Now I'm restarting Obsidian and when I start it back here is what I see: the second button is visually disabled, but all the notes with outgoing links are visible back:

<img width="449" alt="Снимок экрана 2023-08-26 в 18 56 00" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/321d04bc-8fe2-4692-909f-1fbfef5e9714">

And now to make them disappear I should disable first button (not the second!):

<img width="449" alt="Снимок экрана 2023-08-26 в 18 56 45" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/d9a9f2aa-a663-4275-9d7b-28c396336c01">

Looks like a bug to me.

And a quick question: what exactly `Hide indirect notes` toggle does? On my vault this toggle has no effect.

**Comment by vrtmrz** [comment]
Thank you for your patiently reviewing!
Just for some simple bugs, operations could not reflected immediately. The indirectly referenced notes also had not been handled fully.
At v0.17.3, these things should be got better. Would you please try that once? (Nonsensely heavy usage of the CPU is also fixed).

**Comment by VKondakoff** [comment]
Yes, heavy CPU usage is fixed, thanks!

But there is still something wrong with Incoming/Outgoing links toggles.
For example, I have a note `Lillet` with one and only outgoing link to `Drinks`. In v0.17.3 this note is displayed as virtual folder independently of the enabled toggles:

<img width="1062" alt="Снимок экрана 2023-08-28 в 13 01 59" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/d176cb3f-ab39-460d-bb6d-8378d64d5a04">

<img width="1062" alt="Снимок экрана 2023-08-28 в 13 03 35" src="https://github.com/vrtmrz/obsidian-tagfolder/assets/230824/3e456161-26fc-4251-98fe-2fa213bf964f">

Seems like regression from v0.17.2.

**Comment by VKondakoff** [comment]
v0.17.5 starts normally. Thanks!

**Comment by VKondakoff** [comment]
> But there is still something wrong with Incoming/Outgoing links toggles.

This issue is fixed in 0.17.5. Bravo! 👏

**Comment by vrtmrz** [comment]
Thank you very much for the review and I am sorry to postpone the reply!
Since 0.17.2, we have been dealing with performance issues, but we found that we could not use the smart way in some edge cases. Therefore, at 0.17.4, I have tried to search for items in a straightforward but robust way and to streamline their reflection through differential detection. However, due to my inadvertent mistake, every change was detected as a change, which made it very slow.

The mistake has been corrected in 0.17.5! All changes should be properly compared.

> This issue is fixed in 0.17.5. Bravo! 👏

And also, of course, I have been very relieved to hear this! (Unusable behaviour is meaningless, even if works smoothly). Thank you for your kind words.
